common/src/main/java/com/linecorp/centraldogma/common/PathPatternBuilder.java

```diff
@@ -0,0 +1,137 @@
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.centraldogma.common.PathPatternOptions.PathPatternOptionFactory;
+
+/**
+ * Builds a new {@link PathPattern}.
+ *
+ * <h2>Example</h2>
+ * <pre>{@code
+ * final PathPattern factory =
+ *         PathPattern.builder()
+ *                    .startsWith("/foo/bar")
+ *                    .contains("/ext")
+ *                    .extension("json")
+ *                    .build();
+ * }</pre>
+ */
+public final class PathPatternBuilder {
+
+    private final Map<String, PathPatternOption> options = new LinkedHashMap<>();
+
+    /**
+     * Adds {@link PathPatternOptions#ENDS_WITH_OPTION}.
+     */
+    public PathPatternBuilder endsWith(String filename) {
+        put(PathPatternOptions.ENDS_WITH_OPTION, filename);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#EXTENSION_OPTION}.
+     */
+    public PathPatternBuilder extension(String extension) {
+        put(PathPatternOptions.EXTENSION_OPTION, extension);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#STARTS_WITH_OPTION}.
+     */
+    public PathPatternBuilder startsWith(String dirPath) {
+        put(PathPatternOptions.STARTS_WITH_OPTION, dirPath);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#CONTAINS_OPTION}.
+     */
+    public PathPatternBuilder contains(String dirPath) {
+        put(PathPatternOptions.CONTAINS_OPTION, dirPath);
+        return this;
+    }
+
+    /**
+     * Combine 2 patterns into one.
+     */
+    private static String combine(String left, String right) {
+        checkArgument(left.endsWith("/**"), "left should end with \"/**\"");
+        checkArgument(right.startsWith("/**/"), "right should start with \"/**/\"");
+        return left + right.substring(3);
+    }
+
+    /**
+     * Compose one pathPattern from a list of {@code patterns}.
+     */
+    private static String combine(Iterable<PathPattern> patterns) {
+        final Iterator<PathPattern> iter = patterns.iterator();
+        String combinedPattern = iter.next().patternString();
+        while (iter.hasNext()) {
+            combinedPattern = combine(combinedPattern, iter.next().patternString());
+        }
+        return combinedPattern;
+    }
+
+    /**
+     * Returns a newly-created {@link PathPattern} based on the options of this builder.
+     */
+    public PathPattern build() {
+        checkArgument(!options.isEmpty(), "Requires at least one pattern to build in PathPatternBuilder");
+
+        if (options.size() == 1) {
+            return options.entrySet().iterator().next().getValue().getPathPattern();
+        }
+        // given the same precedence, option added after overrides previous one
+        final Map<Integer, PathPatternOption> optionByPrecedence = options.values()
```

Comment:
Hi, for this comment https://github.com/line/centraldogma/pull/809/files#r1115291121 :
 the reason I use `Integer` instead of primitive` int` for precedence is because I use Map to group options by precedence. If there are `contains` and `endsWith`, only one is allowed --> this part will select only one of the two based on which is added by user.

Reply:
I see. However, `PathPatternOption.getPrecedence()` already returns `int`.

Reply:
Noted! Will fix `PathPatternOption.getPrecedence()`.